### DISCUSSION TITLE
「配列の個数以上の指定」について

### DISCUSSION THREAD

**Discussion by ShunHigu** [lit-iphone-course/lit-iphone-course#69] [Question-original]
１　ボタンを押すごとに乱数を発生させ、その数にあった画像を画面上に出し続けたいです。

２　以前も質問させて頂いたのですが、ボタンで出てくる画像一つ一つを配列で表現した結果、次のようなエラーが出てしまいました。配列で宣言する前は問題なく実行できました。

Thread 1: Fatal error: Index out of range

stonenumやenemynumとはボタンを押し乱数を生成させ、if文で振り分けられるごとにstonenumを1ずつ増やし、画面上に何個の画像があるか分かるようにするためです。

３　エラー文を調べた結果、配列の範囲に収まっていないと書かれていました。

４　ImageStoneの宣言時にvar ImageStone:[UIImageView]=[]のように宣言したことに間違いがあるかもと考えました。
もしくは、stonenumの数え方に間違いがあるかもとも考えました。

var StoneImage:[UIImageView]=[]
    var EnemyImage:[UIImageView]=[]
    var SekkiImage:[UIImageView]=[]
    
    var enemynum:Int=0
    var stonenum:Int=0
    
    @IBAction func Renda(){
        num=Int.random(in: 1...100)
        let xrand=Double.random(in: 40...370)
        let yrand=Double.random(in: 40...450)
      
        if num<=75{
            let image1:UIImage = UIImage(named:"Stone")!
            StoneImage[stonenum]=UIImageView(image:image1)
            StoneImage[stonenum].center = CGPoint(x:xrand, y:yrand)
            view.addSubview(StoneImage[stonenum])
            stonenum=stonenum+1
        }
        if (num>=76)&&(num<=95){
            let image1:UIImage = UIImage(named:"Stone")!
            EnemyImage[enemynum]=UIImageView(image:image1)
            EnemyImage[enemynum].center = CGPoint(x:xrand, y:yrand)
            view.addSubview(EnemyImage[enemynum])
            enemynum=enemynum+1
        }
        if num>=96{
            let image1:UIImage = UIImage(named:"Timer")!
            let TimerImage=UIImageView(image:image1)
            TimerImage.center = CGPoint(x:xrand, y:yrand)
            view.addSubview(TimerImage)
        }
    }

５　https://github.com/ShunHigu/OriginalNew
修正内容を「修正１」にcommit&pushしました。

**Comment by Yonpachi48** [comment]
`StoneImage[stonenum]=UIImageView(image:image1)`ではなく`StoneImage.append(UIImageView(image:image1))`とすることで正しく動きます。
違いは、前者が**代入**で後者が**追加**です。

`Index out of range`とは配列の中の要素を超えてしまっているため起こるエラーです。
今回だと配列の宣言時に`var StoneImage:[UIImageView]=[]`このように空の配列を代入している(要素のないの配列を定義している)ため`StoneImage[stonenum]=UIImageView(image:image1)`(stonenum=0)で存在しないStoneImage[0]に何かを代入しようとした結果、配列の要素数を超えてしまい`Index out of range`となってしまいました。
そこで`StoneImage.append(追加する要素)`とすることで、StoneImageに新たな要素を追加することができるようになります！

**Reply by ShunHigu** [reply]
ありがとうございます！できました！

**Reply by ShunHigu** [reply]
すいません！追加で質問なのですが、大きさを変えるときのコードをCGPointからCGRectに変えたいのですがなかなか上手くいきません！
StoneImage[stonenum]:CGRect=CGRect(x: xrand, y: yrand, width: 100, height: 100)
別のやり方があれば教えていただきたいです。
